src/routers/alpha-router/gas-models/tick-based-heuristic-gas-model.ts

```diff
@@ -0,0 +1,303 @@
+                gasCostInTermsOfAmountToken.toExact(),
+              executionPrice: executionPrice.toSignificant(6),
+              syntheticGasCostInTermsOfQuoteToken:
+                syntheticGasCostInTermsOfQuoteToken.toSignificant(6),
+            },
+            'New gasCostInTermsOfQuoteToken calculated with synthetic quote token price is less than original'
+          );
+
+          gasCostInTermsOfQuoteToken = syntheticGasCostInTermsOfQuoteToken;
+        }
+      }
+
+      // If gasCostInTermsOfQuoteToken is null, both attempts to calculate gasCostInTermsOfQuoteToken failed (nativePool and amountNativePool)
+      if (gasCostInTermsOfQuoteToken === null) {
+        log.info(
+          `Unable to find ${nativeCurrency.symbol} pool with the quote token, ${quoteToken.symbol}, or amount Token, ${amountToken.symbol} to produce gas adjusted costs. Route will not account for gas.`
+        );
+        return {
+          gasEstimate: baseGasUse,
+          gasCostInToken: CurrencyAmount.fromRawAmount(quoteToken, 0),
+          gasCostInUSD: CurrencyAmount.fromRawAmount(usdToken, 0),
+        };
+      }
+
+      return {
+        gasEstimate: baseGasUse,
+        gasCostInToken: gasCostInTermsOfQuoteToken,
+        gasCostInUSD: gasCostInTermsOfUSD!,
+        gasCostInGasToken: gasCostInTermsOfGasToken,
+      };
+    };
+
+    return {
+      estimateGasCost: estimateGasCost.bind(this),
+      calculateL1GasFees,
+    };
+  }
+
+  protected estimateGas(
+    routeWithValidQuote: TRouteWithValidQuote,
+    gasPriceWei: BigNumber,
+    chainId: ChainId,
+    providerConfig?: GasModelProviderConfig
+  ) {
+    const totalInitializedTicksCrossed = BigNumber.from(
+      Math.max(1, _.sum(routeWithValidQuote.initializedTicksCrossedList))
+    );
+    const totalHops = BigNumber.from(routeWithValidQuote.route.pools.length);
+
+    let hopsGasUse = COST_PER_HOP(chainId).mul(totalHops);
+
+    // We have observed that this algorithm tends to underestimate single hop swaps.
+    // We add a buffer in the case of a single hop swap.
+    if (totalHops.eq(1)) {
+      hopsGasUse = hopsGasUse.add(SINGLE_HOP_OVERHEAD(chainId));
+    }
+
+    // Some tokens have extremely expensive transferFrom functions, which causes
+    // us to underestimate them by a large amount. For known tokens, we apply an
+    // adjustment.
+    const tokenOverhead = TOKEN_OVERHEAD(chainId, routeWithValidQuote.route);
+
+    const tickGasUse = COST_PER_INIT_TICK(chainId).mul(
+      totalInitializedTicksCrossed
+    );
+    const uninitializedTickGasUse = COST_PER_UNINIT_TICK.mul(0);
+
+    /*
+    // Eventually we can just use the quoter gas estimate for the base gas use
+    // It will be more accurate than doing the offchain gas estimate like below
+    // It will become more critical when we are going to support v4 hookful routing,
+    // where we have no idea how much gas the hook(s) will cost.
+    // const baseGasUse = routeWithValidQuote.quoterGasEstimate
```

Comment:
@cgkol @xrsv basically I think in long term, we can replace all below gas heuristics:

```
    const totalInitializedTicksCrossed = BigNumber.from(
      Math.max(1, _.sum(routeWithValidQuote.initializedTicksCrossedList))
    );
    const totalHops = BigNumber.from(routeWithValidQuote.route.pools.length);
    let hopsGasUse = COST_PER_HOP(chainId).mul(totalHops);
    // We have observed that this algorithm tends to underestimate single hop swaps.
    // We add a buffer in the case of a single hop swap.
    if (totalHops.eq(1)) {
      hopsGasUse = hopsGasUse.add(SINGLE_HOP_OVERHEAD(chainId));
    }

    const baseGasUse = BASE_SWAP_COST(chainId)
      .add(hopsGasUse)
      .add(tokenOverhead)
      .add(tickGasUse)
      .add(uninitializedTickGasUse)
      .add(providerConfig?.additionalGasOverhead ?? BigNumber.from(0));
    const baseGasCostWei = gasPriceWei.mul(baseGasUse);
```

with  ```const baseGasUse = routeWithValidQuote.quoterGasEstimate ```, and it will actually be more accurate. 

And v3 [QuoterV2](https://github.com/Uniswap/v3-periphery/blob/main/contracts/lens/QuoterV2.sol), v4 [Quoter](https://github.com/Uniswap/v4-periphery/blob/main/src/lens/Quoter.sol), and mixed [MixedRouteQuoterV1](https://github.com/Uniswap/swap-router-contracts/blob/main/contracts/lens/MixedRouteQuoterV1.sol) can return a much more accurate `gasEstimate` then calculating here in SOR

Reply:
shall we add a ticket for this? it does indeed sound like we should use it; any reason why we havent so far?

Reply:
Not sure, but let me log a ticket to keep track.

Reply:
https://linear.app/uniswap/issue/ROUTE-245/tech-debt-sor-gas-model-should-use-the-gas-estimate-returned-from yep added